### ISSUE TITLE
Add ability to do phased roll-outs of configs

### ISSUE THREAD

**Issue by blackfist** [blackfist/windmill#19]
It would be super cool if I wanted to roll out some new queries if I could release it to some percentage of the servers randomly and watch for performance problems before full release. No idea how to make that happen yet.

**Comment by blackfist** [comment]
Here is some pseudocode around what I was thinking this morning.

Let's say you have a config file web.conf and you want to test something new. Create a file called web_test_25.conf where 25 means you want 25% of endpoints to run that configuration.

```
POST request comes in to /api/config/web

validate node secret

Find all the files that start with "web".
If there's only one, then send it and record that as the last config sent. end.
If there are two, then check if the last config sent was "web_test_25". If so, send that and record that as the last config sent. end.

# If we get here it means that there is a rollout going on and this endpoint might need the new config

get count of endpoints that have "web" config
get count of endpoints that have "web_test_25" config

if count("web_test_25") >= count("web_test_25") + count("web") then send "web" and record that as the last config sent.

Else, random roll with 25% chance of success
If fail, send "web" and record that as the last config sent
if success send "web_test_25" and record that as the last config sent
```

**Comment by blackfist** [comment]
My concern with the pseudocode above is that in a large deployment if you're rolling a new configuration out to 10% of your clients you're going to have a lot of database round trips. 90% of the time you're going to be querying the node secret, counting end points with the base config, counting endpoints with the new config, and recording which config was last sent.

**Comment by blackfist** [comment]
gem install bernoulli

```
require 'bernoulli/shorthand'

probability_distribution = binomdist(1,0.25)

if probability_distribution.sample[0] == 1
  # do something
else
  # do something else
```

**Comment by blackfist** [comment]
The word on the street is that Rendezvous hashing might be the right way to do this. I'm just recording that here so I don't forget.

**Comment by blackfist** [comment]
Closed in 125ee4758bee0b4f59f4243675eb65307c255d9f
